### PR TITLE
Replace 'pm list packages' with 'dumpsys package'

### PR DESCRIPTION
**Pull request by SecTheTech**
In some cases, the command 'pm list packages' will not list all enabled third party packages (same for -e and -s). 
https://stackoverflow.com/questions/63416599/adb-shell-pm-list-packages-missing-some-packages

### CODE REVIEW

File: adbe/adb_enhanced.py [hunk 1]
```diff
@@ -432,16 +435,18 @@ def _get_top_activity_data():
...
         line = line.strip()
-        regex_result = re.search(r'ActivityRecord{.* (\S+)/(\S+)', line)
-        if regex_result is None:
-            continue
-        app_name, activity_name = regex_result.group(1), regex_result.group(2)
-        # If activity name is a short hand then complete it.
-        if activity_name.startswith('.'):
-            activity_name = '%s%s' %(app_name, activity_name)
-        return app_name, activity_name
+        if line.startswith('mFocusedApp'):
```
Comment (ashishb) [inline adbe/adb_enhanced.py:438]: Is this change intentional? I think this is the exact code that I removed in https://github.com/ashishb/adb-enhanced/commit/69bb243f9d4508279554d074ed93ddc8e01100c0

File: adbe/adb_enhanced.py [hunk 2]
```diff
@@ -922,9 +927,13 @@ def _get_all_packages(pm_cmd):
 
 
 def list_all_apps():
-    cmd = 'pm list packages'
-    packages = _get_all_packages(cmd)
-    print('\n'.join(packages))
+    # The command "pm list packages -e" does not return all installed and enabled Apps, the alternative is to use
```
Comment (ashishb) [inline adbe/adb_enhanced.py:930]: So what exactly does `pm list packages -e` miss? Disabled apps?

File: adbe/adb_enhanced.py [hunk 3]
```diff
@@ -1562,43 +1579,3 @@ def switch_screen(switch_type):
                                  "screen control operation. Error: %s" % e)
 
     return o
-
```
Comment (ashishb) [inline adbe/adb_enhanced.py:1565]: Please rebase your code to avoid reverting this addition.

File: adbe/adb_enhanced.py [hunk 4]
```diff
@@ -16,6 +16,7 @@
 if sys.version_info >= (3, 5):
     try:
         import asyncio_helper
+
```
Comment (ashishb) [inline adbe/adb_enhanced.py:19]: Considering removing such spurious changes to minimize code diff.

File: adbe/adb_enhanced.py [hunk 5]
```diff
@@ -922,9 +927,13 @@ def _get_all_packages(pm_cmd):
 
 
 def list_all_apps():
-    cmd = 'pm list packages'
-    packages = _get_all_packages(cmd)
-    print('\n'.join(packages))
+    # The command "pm list packages -e" does not return all installed and enabled Apps, the alternative is to use
+    # "dumpsys package" instead.
+    cmd = 'dumpsys package'
```
Comment (ashishb) [inline adbe/adb_enhanced.py:932]: Looks good. Please add a sample output in the conversations on this PR

**Review by ashishb — COMMENTED** [review]
Thanks for your contribution.

File: adbe/adb_enhanced.py (hunk 1 above)
Comment (SecTheTech) [reply adbe/adb_enhanced.py:438]: Nope, I forked the master branch.

File: adbe/adb_enhanced.py (hunk 2 above)
Comment (SecTheTech) [reply adbe/adb_enhanced.py:930]: `pm list packages -e` for some unknown reason does not list some enabled apps.  
`-e` should show all enabled Apps including system Apps  which is not the case as discussed in the thread: https://stackoverflow.com/questions/63416599/adb-shell-pm-list-packages-missing-some-packages

File: adbe/adb_enhanced.py (hunk 3 above)
Comment (SecTheTech) [reply adbe/adb_enhanced.py:1565]: Sure. Sorry for that!

**Comment by SecTheTech** [comment]
Command output:
```
$ adbe apps list all
com.android.cts.priv.ctsshim
com.android.internal.display.cutout.emulation.corner
com.android.internal.display.cutout.emulation.double
com.android.providers.telephony
com.android.dynsystem
com.android.providers.calendar
com.android.providers.media
com.android.theme.icon.square
com.android.internal.systemui.navbar.gestural_wide_back
....
```

File: adbe/adb_enhanced.py (hunk 2 above)
Comment (ashishb) [reply adbe/adb_enhanced.py:924]: Thanks. It is fine if we don't know the underlying cause. Add this StackOverflow thread in the code comments as well.

File: adbe/adb_enhanced.py (hunk 1 above)
Comment (ashishb) [reply adbe/adb_enhanced.py]: Just do a `git pull origin master` in that case. Your branch has become stale. I changed this code <24 hours ago.

File: adbe/adb_enhanced.py (hunk 3 above)
Comment (ashishb) [reply adbe/adb_enhanced.py]: No worries.

**Review by ashishb — APPROVED** [review]
Approved pending the addition of StackOverflow thread as a comment for future reference.

File: adbe/adb_enhanced.py (hunk 2 above)
Comment (SecTheTech) [reply adbe/adb_enhanced.py:924]: I saw that the regex is reported as an anomaly by DeepSource Python, Can we replace it by this one `dumpsys package | grep "Package \[" | cut -d "\[" -f2 | cut -d "\]" -f1`.

Comment (ashishb) [reply adbe/adb_enhanced.py:924]: The new code is fine. Just confirm that `cut` command's arguments are the same on both Linux and Mac.

Comment (SecTheTech) [reply adbe/adb_enhanced.py:924]: It should be the same (https://superuser.com/questions/179368/how-does-macs-command-line-compare-to-linux), unfortunately I do not have a Mac to test. Can someone else confirm that please ?

File: adbe/adb_enhanced.py [hunk 6]
```diff
@@ -13,33 +13,33 @@
...
+    try :
         import asyncio_helper
+
         _ASYNCIO_AVAILABLE = True
-    except ImportError:
+    except ImportError :
         # This is to deal with python versions below 3.5
         _ASYNCIO_AVAILABLE = False
-else:
+else :
```
[the remark this answers is not in the document]
Comment (ashishb) [reply adbe/adb_enhanced.py]: Please revert

File: adbe/adb_enhanced.py [hunk 7]
```diff
@@ -13,33 +13,33 @@
...
 # asyncio was introduced in version 3.5
-if sys.version_info >= (3, 5):
-    try:
+if sys.version_info >= (3, 5) :
+    try :
         import asyncio_helper
+
         _ASYNCIO_AVAILABLE = True
-    except ImportError:
+    except ImportError :
```
[the remark this answers is not in the document]
Comment (ashishb) [reply adbe/adb_enhanced.py]: Please revert

File: adbe/adb_enhanced.py (hunk 2 above)
Comment (ashishb) [reply adbe/adb_enhanced.py:924]: I got `cut: bad delimiter` as an error on Mac OS. How about we do this parsing in Python instead? That's much more portable as well.

Comment (SecTheTech) [reply adbe/adb_enhanced.py:924]: Yes sure, I will update it.

**Comment by ashishb** [comment]
So, I checked `dumpsys package | grep "Package \["` today. It includes some duplicates but even after removing all duplicates I saw 420 packages as opposed to 403 packages from `adb shell pm list packages`, I do see some weird and interesting services listed. So, yeah, let's merge this

File: adbe/adb_enhanced.py [hunk 8]
```diff
@@ -921,10 +921,15 @@ def _get_all_packages(pm_cmd):
...
 def list_all_apps():
-    cmd = 'pm list packages'
-    packages = _get_all_packages(cmd)
-    print('\n'.join(packages))
+    cmd = 'dumpsys package'
+    pattern_packages = re.compile('Package \[(.*?)\]')
+    return_code, result, _ = execute_adb_shell_command2(cmd)
+    if return_code != 0:
+        print_error_and_exit('Command "%s" failed, something is wrong' % cmd)
+    all_apps = re.findall(pattern_packages, result)
```
Comment (ashishb) [inline adbe/adb_enhanced.py:931]: This is not printing the results. I will fix that in https://github.com/ashishb/adb-enhanced/pull/135
